pygfx/objects/_ruler.py

```diff
@@ -0,0 +1,489 @@
+        self._min_tick_dist = value
+
+    @property
+    def ticks_at_end_points(self):
+        """Whether to show tickmarks at the end-points.
+
+        You must call `set_ticks()` for this change to take effect.
+        """
+        return self._ticks_at_end_points
+
+    @ticks_at_end_points.setter
+    def ticks_at_end_points(self, value):
+        self._ticks_at_end_points = bool(value)
+
+    # -- Methods
+
+    def update(self, camera, canvas_size):
+        """Update the ruler.
+
+        This must be called on every draw, right before rendering.
+
+        Returns a dictionary with the following fields:
+
+        * "tick_step": the calculated auto-tick-step.
+        * "tick_values": the tick values that will be shown.
+        """
+
+        # Determine which part of the ruler is on screen and its length in screen pixels
+        self._configure_for_screen(camera, canvas_size)
+
+        # Anchor
+        screen_vec = self._visible_part_screen_vec
+        self._calculate_text_anchor(np.arctan2(screen_vec[1], screen_vec[0]))
+
+        # Get the dict with visible ticks
+        tick_auto_step = self._calculate_tick_step()
+        visible_ticks = self._get_ticks_dict(tick_auto_step)
+
+        # Update objects to show these ticks
+        self._update_sub_objects(visible_ticks, tick_auto_step)
+
+        # Return stats. This is a dict, so we can add more stuff later, if needed.
+        return {
+            "tick_step": tick_auto_step,
+            "tick_values": list(visible_ticks.keys()),
+        }
+
+    def _configure_for_screen(self, camera, canvas_size):
+        """Make the ruler aware of the camera and viewport size."""
+
+        # Yuk, but needed
+        camera.update_projection_matrix()
+
+        half_canvas_size = 0.5 * np.array(canvas_size, np.float64)
+
+        # Get ndc coords for begin and end pos
+        positions = np.column_stack(
+            [
+                np.row_stack([self._start_pos, self._end_pos]),
+                np.ones((2, 1), np.float64),
+            ]
+        )
+        ndc1, ndc2 = (camera.camera_matrix @ positions[..., None]).reshape(-1, 4)
+
+        ndc1_2d = ndc1[:2] / ndc1[3]
+        ndc2_2d = ndc2[:2] / ndc2[3]
+        screen1 = ndc1_2d * half_canvas_size
+        screen2 = ndc2_2d * half_canvas_size
+
+        # Get what part of the line is visible
+        t1, t2 = get_visible_part_of_line_ndc(ndc1, ndc2)
+
+        # Get corresponding screen coordinates.
+        # Fall back to full line when the selected region is empty,
+        # so that we can still calculate the step size, because calling code
+        # may still need it, e.g. to configure a grid.
+        ndc_t1 = ndc1 * (1 - t1) + ndc2 * t1
+        ndc_t2 = ndc1 * (1 - t2) + ndc2 * t2
+        ndc_t1_2d = ndc_t1[:2] / ndc_t1[3]
+        ndc_t2_2d = ndc_t2[:2] / ndc_t2[3]
+        screen_t1 = ndc_t1_2d * half_canvas_size
+        screen_t2 = ndc_t2_2d * half_canvas_size
```

Comment:
if ndc1 and ndc2 are rows in a single matrix, then you should be able to use broadcasting to have half of the lines of code here right?

Reply:
Nice, will do!